SwiftCompilerSources/Sources/Optimizer/Utilities/WalkUtils.swift

```diff
@@ -0,0 +1,376 @@
+//===--- WalkUtils.swift - Utilities for use-def def-use walks ------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+func TODO(_ message: String = "Unhandled TODO", fun: String = #function,
+          file: String = #file, line: Int = #line) -> Never {
+  fatalError("\(#function)\(file):\(line) \(fun) \(message)")
+}
+
+func FATAL(_ message: String = "", fun: String = #function,
+          file: String = #file, line: Int = #line) -> Never {
+  fatalError("\(#function)\(file):\(line) \(fun) \(message)")
+}
+
+public enum WalkResult {
+  /// skip traversal
+  case ignore
+  /// stop the walk, add the current use/def and path to the result
+  case abortWalk
+  /// continue the walk
+  case continueWalk
+}
+
+protocol MemoizedUp {
+  mutating func shouldRecomputeUp(value def: Value, path: SmallProjectionPath) -> SmallProjectionPath?
+}
+
+protocol MemoizedDown {
+  mutating func shouldRecomputeDown(value def: Value, path: SmallProjectionPath) -> SmallProjectionPath?
+}
+
+protocol ValueDefUseWalker : MemoizedDown {
+  mutating
+  func visitUse(ofValue operand: Operand, path: SmallProjectionPath, isLeaf: Bool) -> WalkResult
+}
+
+extension ValueDefUseWalker {
+  private mutating
+  func visitAndWalkDown(_ operand: Operand, path: SmallProjectionPath, values: Instruction.Results,
+                        newPath: SmallProjectionPath) -> Bool {
+    let maybeContinue = visitUse(ofValue: operand, path: path, isLeaf: false)
+    switch maybeContinue {
+    case .continueWalk:
+      for result in values {
+        // TODO: insert shouldRecomputeDown call?
+        // EscapeInfo doesn't
+        if walkDown(value: result, path: newPath) {
+          return true
+        }
+      }
+      return false
+    default:
+      return maybeContinue == .abortWalk
+    }
+  }
+  
+  /// Note: keep it symmetric to `ValueUseDefWalker.walkUp`
+  mutating
+  func walkDown(value def: Value, path: SmallProjectionPath) -> Bool {
+    for operand in def.uses {
+      if operand.isTypeDependent { continue }
+      
+      // .none              : current not visited
+      // .some(let aborted) : current and all recursive paths visited and `aborted = visitResult is .abortWalk`
+      var visitAborted: Bool?
```

Comment:
Please no optional booleans. If you need such a thing, define a new enum.
But anyway, if you inline the `resultPath` functions, you don't need it here at all.